File: solutions/haiku-1/haiku/ruby/brackets.rb
while line = STDIN.gets
  line = line.chomp

  stack = []
  balanced = true

  # Map closing brackets to their opening counterparts
  matches = { ')' => '(', ']' => '[', '}' => '{' }

  line.each_char do |char|
    if '([{'.include?(char)
      # Push opening brackets onto stack
      stack.push(char)
    elsif matches.key?(char)
      # Check if closing bracket matches top of stack
      if stack.empty? || stack[-1] != matches[char]
        balanced = false
        break
      end
      stack.pop
    end
  end

  # Stack must be empty for balanced brackets
  balanced = false unless stack.empty?

  puts balanced ? 'yes' : 'no'
end

File: solutions/haiku-1/haiku/perl/brackets.pl
#!/usr/bin/perl
use strict;
use warnings;

while (my $line = <>) {
    chomp($line);

    my @stack;
    my $valid = 1;
    my %closing = ('(' => ')', '[' => ']', '{' => '}');

    for my $char (split //, $line) {
        if ($char =~ /[(\[{]/) {
            # Opening bracket - push to stack
            push @stack, $char;
        } elsif ($char =~ /[)\]}]/) {
            # Closing bracket
            if (!@stack) {
                # No matching opening bracket
                $valid = 0;
                last;
            }

            my $open = pop @stack;

            # Check if closing bracket matches opening bracket
            if ($closing{$open} ne $char) {
                $valid = 0;
                last;
            }
        }
    }

    # Valid only if no errors and stack is empty
    if ($valid && !@stack) {
        print "yes\n";
    } else {
        print "no\n";
    }
}
